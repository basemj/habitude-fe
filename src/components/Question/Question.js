import React, {Component} from 'react';
import Prism from 'prismjs';

import '../../assets/css/prism.css';

export default class Question extends Component {

  sendAnswer = (answer) => {
    if (this.props.answerClickHandler) {
      this.props.answerClickHandler(answer);
    }
  }
  
  componentDidMount() {
    Prism.highlightAll();
  }
  
  render() {
    if (!this.props.question && !this.props.answers) {
      return null;
    }

    const {
      title,
      description,
      codeSnippet,
      codeSnippetLang,
      answers
    } = this.props.question;  

    return (
      <div>
        <h1>{title ? title : 'Question'}</h1>
        {description ? <p>{description}</p> : null}

        {
          codeSnippet ?
          <pre>
            <code className={codeSnippetLang ? `language-${codeSnippetLang}` : `language-js`}>
            {`${codeSnippet}`}
            </code>
          </pre> :
          null
        }

        <ul>
          {answers.map((answer) => 
            <li key={answer.id} onClick={this.sendAnswer.bind(this, answer)}>{answer.text}</li>
          )}
        </ul>
      </div>
    )
  }
}
