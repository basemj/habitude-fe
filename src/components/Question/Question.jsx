import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Prism from 'prismjs';

import '../../assets/css/prism.css';

export default class Question extends Component {
  componentDidMount() {
    Prism.highlightAll();
  }

  render() {
    const {
      answerClickHandler,
      question,
    } = this.props;

    const {
      title,
      description,
      codeSnippet,
      codeSnippetLang,
      answers,
    } = question;

    const sendAnswer = (answer) => {
      if (answerClickHandler) {
        answerClickHandler(answer);
      }
    };

    if (!question || !answers) {
      return null;
    }

    return (
      <div>
        <h1>{title}</h1>
        {description ? <p>{description}</p> : null}

        {
          codeSnippet
            ? (
              <pre>
                <code className={`language-${codeSnippetLang}`}>
                  {`${codeSnippet}`}
                </code>
              </pre>
            )
            : null
        }

        <ul>
          {answers.map((answer) => (
            <li key={answer.id}>
              <button type="button" onClick={sendAnswer.bind(this, answer)} onKeyDown={sendAnswer.bind(this, answer)}>{answer.text}</button>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

Question.propTypes = {
  answerClickHandler: PropTypes.func.isRequired,
  question: PropTypes.exact({
    title: PropTypes.string,
    description: PropTypes.string,
    codeSnippet: PropTypes.string,
    codeSnippetLang: PropTypes.string,
    answers: PropTypes.arrayOf(PropTypes.object),
  }),
};

Question.defaultProps = {
  question: {
    title: 'Question',
    codeSnippetLang: 'js',
    answers: [],
  },
};
