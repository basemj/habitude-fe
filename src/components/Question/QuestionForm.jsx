import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Prism from 'prismjs';
import styles from './QuestionForm.module.css';
import '../../assets/css/prism.css';
import AnswerForm from '../Answer/AnswerEditForm';

const QuestionForm = (
  {
    question: {
      id, description: initialDescription, script: initialScript, answers: initialAnswers,
    },
    onSubmit,
  },
) => {
  const [description, setDescription] = useState(initialDescription);
  const [script, setScript] = useState(initialScript);
  const [answers, setAnswer] = useState(initialAnswers);
  const [answersHasChanged, setAnswerChangeStatus] = useState(false);

  useEffect(() => Prism.highlightAll());

  useEffect(() => {
    const timer = setTimeout(() => {
      setAnswerChangeStatus(false);
    }, 3000);
    return () => clearTimeout(timer);
  }, [answersHasChanged]);

  const saveAnswer = (answer, index) => {
    const i = index === undefined ? answers.length : index;
    answers.splice(i, 1, answer);
    setAnswer([...answers]);
    setAnswerChangeStatus(true);
  };

  const submit = (e) => {
    e.preventDefault();
    const data = {
      id,
      description,
      script,
      answers,
    };
    onSubmit(data);
  };

  return (
    <div>
      <h1>Edit question</h1>
      {answersHasChanged ? <div className={styles.answersHasChanged}>Answer has changed but the question was not saved!</div> : ''}
      <form className={styles.form} onSubmit={submit}>
        <div>
          <label htmlFor="description">
            Description
            <textarea id="description" name="description" onChange={(e) => setDescription(e.target.value)} value={description} />
          </label>
        </div>
        <div>
          <label htmlFor="script">
            Script
            <textarea id="script" name="script" onChange={(e) => setScript(e.target.value)} value={script} />
          </label>
        </div>
        <div>
          <pre>
            <code className="language-js">{script}</code>
          </pre>
        </div>
        <div>
          <button type="submit">Save</button>
        </div>
      </form>
      <p>
        {'There are '}
        {answers.length}
        {' answers added.'}
      </p>
      {answers.map((answer, index) => <AnswerForm key={`answer-${Math.random() + index}`} answer={answer} onSubmit={(a) => saveAnswer(a, index)} />)}
      <AnswerForm onSubmit={(a) => saveAnswer(a)} />

    </div>
  );
};

QuestionForm.propTypes = {
  question: PropTypes.shape({
    id: PropTypes.string,
    description: PropTypes.string,
    script: PropTypes.string,
    answers: PropTypes.array,
  }),
  onSubmit: PropTypes.func,
};
QuestionForm.defaultProps = {
  question: {
    id: '',
    answers: [],
    description: '',
    script: '',
  },
  onSubmit: () => {},
};

export default QuestionForm;
