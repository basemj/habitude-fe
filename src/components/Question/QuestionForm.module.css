.form {

}
.form textarea {
  width: 100%;
  min-height: 150px;
}
.answersHasChanged {
  position: fixed;
  background-color: grey;
  color: white;
  padding: 30px;
  top: 0;
  right: 0;
}
