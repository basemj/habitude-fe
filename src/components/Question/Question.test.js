import React from 'react';
import { mount } from 'enzyme';
import Question from './Question';

const questionObject = {
  title: 'Simple question',
  description: 'Here is the description for this simple question.',
  codeSnippet: '// foo bar\nvar foo = \'bar\';\n\nfunction hello() {\n\treturn \'Hello World!\'\n}',
  codeSnippetLang: 'js',
  answers: [
    {
      id: 'answer1',
      text: 'This is a sample answer',
    },
    {
      id: 'answer2',
      text: 'This could be the right answer',
    },
    {
      id: 'answer3',
      text: 'Here\'s another answer',
    },
  ],
};

describe('Question', () => {
  describe('snapshot', () => {
    let component;
    it('it renders', () => {
      component = mount(<Question />);
      expect(component.html()).toMatchSnapshot();
    });

    it('renders title in a H1', () => {
      component = mount(<Question question={questionObject} />);
      expect(component.find('h1').text()).toBe(questionObject.title);
    });

    it('renders question description', () => {
      component = mount(<Question question={questionObject} />);
      expect(component.find('p').text()).toBe(questionObject.description);
    });

    it('renders code snippet with correct language', () => {
      component = mount(<Question question={questionObject} />);
      expect(component.find('code').text()).toBe(questionObject.codeSnippet);
      expect(component.find('code').hasClass(`language-${questionObject.codeSnippetLang}`)).toBe(true);
    });

    it('renders all three answers', () => {
      component = mount(<Question question={questionObject} />);
      expect(component.find('li').length).toBe(3);
    });
  });

  describe('action', () => {
    let component;
    const answerClickHandler = jest.fn();

    it('invokes a callback with the correct data', () => {
      component = mount(<Question question={questionObject} answerClickHandler={answerClickHandler} />);

      expect(answerClickHandler).not.toHaveBeenCalled();
      component.find('li').at(1).simulate('click');
      expect(answerClickHandler).toHaveBeenCalled();
      expect(answerClickHandler).toHaveBeenCalledWith(questionObject.answers[1]);
    });
  });
});
