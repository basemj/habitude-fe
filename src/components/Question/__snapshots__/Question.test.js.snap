// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question snapshot it renders 1`] = `null`;
